pkg/worker/worker.go

```diff
@@ -0,0 +1,300 @@
+		return ErrJobNotFound
+	}
+	if job.Status == StatusStopped {
+		return ErrJobAlreadyStopped
+	}
+	if job.Status == StatusFinished {
+		return ErrJobAlreadyFinished
+	}
+
+	err = syscall.Kill(job.PID, syscall.SIGKILL)
+	if err != nil {
+		return errors.Wrapf(err, "error killing the process with pid: %d", job.PID)
+	}
+
+	job.Status = StatusStopped
+	job.StoppedAt = time.Now()
+	err = job.store()
+
+	return err
+}
+
+// Query returns information about the job.
+func (worker *Worker) Query(jobUUID uuid.UUID) (*Job, error) {
+	job, err := worker.DataStorage.Retrieve(jobUUID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error retrieving job with UUID: %v", jobUUID)
+	}
+	if job.UUID == uuid.Nil {
+		return nil, ErrJobNotFound
+	}
+
+	return job, nil
+}
+
+// JobStatus defines a type for job statuses.
+type JobStatus string
+
+// The constants define all possible statuses of the job.
+var (
+	StatusCreated  = JobStatus("created")
+	StatusRunning  = JobStatus("running")
+	StatusError    = JobStatus("error")
+	StatusFinished = JobStatus("finished")
+	StatusStopped  = JobStatus("stopped")
+)
+
+// Job is an abstraction around Linux process that contains logs and metadata.
+type Job struct {
+	DataStorage DataStorage
+	Command     []string
+	Status      JobStatus
+	PID         int
+	UUID        uuid.UUID
+	CreatedAt   time.Time
+	StartedAt   time.Time
+	StoppedAt   time.Time
+	FinishedAt  time.Time
+	User        string
+	rw          sync.RWMutex
+	logs        []byte
+}
+
+// Write allows to add logs to the existing ones in the DataStorage.
+func (job *Job) Write(p []byte) (n int, err error) {
+	job.rw.Lock()
+	job.logs = append(job.logs, p...)
+	job.rw.Unlock()
+	err = job.DataStorage.Store(job)
+	return len(p), err
+}
+
+// Logs returns current logs stored in job.
+func (job *Job) Logs() []byte {
+	job.rw.RLock()
+	defer job.rw.RUnlock()
+	logs := make([]byte, len(job.logs))
+	copy(logs, job.logs)
+	return logs
+}
+
+// SetLogs allows to overwrite logs stored in job.
+func (job *Job) SetLogs(newLogs []byte) {
+	job.rw.Lock()
+	defer job.rw.Unlock()
+	logs := make([]byte, len(newLogs))
+	copy(logs, newLogs)
+	job.logs = logs
+}
+
+// newJob assigns the default, required values returning new job instance.
+func newJob(dataStorage DataStorage, command []string, user string) Job {
+	return Job{
+		DataStorage: dataStorage,
+		Status:      StatusCreated,
+		Command:     command,
+		UUID:        uuid.New(),
+		CreatedAt:   time.Now(),
+		User:        user,
+		logs:        []byte{},
+	}
+}
+
+// start runs the the command in a new process, listens for StdIn and StdErr
+// and logs them to the DataStorage.
+// If the DataStorage is not accessible, the errors are raised up to the Worker.Start
+// method and logged to the standard output.
+// TODO: Add stopping subprocesses of the started processes.
+// TODO: Add separation of the StdIn and StdErr.
+func (job *Job) start() error {
+	cmd := exec.Command(job.Command[0], job.Command[1:]...)
```

Comment:
I'd add a check to make sure that indexing won't lead to a panic.

Reply:
Added some checks.